dev/io.openliberty.org.jboss.resteasy.common/src/org/jboss/resteasy/plugins/server/servlet/ServletUtil.java

```diff
@@ -0,0 +1,179 @@
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package org.jboss.resteasy.plugins.server.servlet;
+
+import org.jboss.resteasy.core.Headers;
+import org.jboss.resteasy.specimpl.ResteasyHttpHeaders;
+import org.jboss.resteasy.specimpl.ResteasyUriInfo;
+import org.jboss.resteasy.specimpl.ResteasyUriInfo.InitData;
+import org.jboss.resteasy.util.HttpHeaderNames;
+import org.jboss.resteasy.util.MediaTypeHelper;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Cookie;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public class ServletUtil
+{
+   private static Map<String, InitData> initDataCache = new HashMap<String, InitData>();      // Liberty change
+    
+   public static ResteasyUriInfo extractUriInfo(HttpServletRequest request, String servletPrefix)
+   {
+      String contextPath = request.getContextPath();
+      if (servletPrefix != null && servletPrefix.length() > 0 && !servletPrefix.equals("/"))
+      {
+         if (!contextPath.endsWith("/") && !servletPrefix.startsWith("/"))
+            contextPath += "/";
+         contextPath += servletPrefix;
+      }
+      String queryString = request.getQueryString();
+      String absolute;
+      if (queryString != null && queryString.length() > 0)
+      {
+         absolute = request.getRequestURL().append('?').append(queryString).toString();
+      }
+      else
+      {
+         absolute = request.getRequestURL().toString();
+      }
+      if (!absolute.contains(contextPath))
+      {
+         String encodedContextPath = Arrays.stream(contextPath.substring(1).split("/"))
+                                           .map(s -> {
+                                              try {
+                                                 return URLEncoder.encode(s, "UTF-8");
+                                                } catch (UnsupportedEncodingException ex) {
+                                                 return s;
+                                              }
+                                             })
+                                           .collect(Collectors.joining("/", "/", ""));
+         if (absolute.contains(encodedContextPath))
+         {
+            absolute = absolute.replace(encodedContextPath, contextPath);
+         }
+      }
+      // Liberty change begin      
+      ResteasyUriInfo myResteasyUriInfo;
+      if(InitData.canBeCached(absolute)) {
+          String cacheKey = absolute + contextPath;
```

Comment:
```suggestion
          String cacheKey = InitData.getCacheKey(absolute, contextPath);
```

Reply:
Nice - done.

Reply:
It turns out that InitData.getCacheKey is fairly expensive - 0.43%-0.48% in two profiles I took with this change. The cost is in the search for the queryString delimiter
`         int queryIdx = absoluteUri.lastIndexOf('?');
`
So instead, let's create the cacheKey using the output of getRequestURL() which is already called in the method, and per the API has the query string params removed already.